docs/programmers_guide/witness_formal_spec.md

```diff
@@ -2,100 +2,263 @@
 
 The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
 
-## The Stack
+## Data Types
 
-Witnesses are executed on a stack machine that builds tries/calculates hashes.
+### Basic
 
-The stack consists of stack items.
-Each item is a pair: `(node, hash)`, where `node` is an object representing the Merkle trie node, and `hash` is 32 byte node hash value. The exact implementation details of `node`s and `hash`es are not the part of the spec.
+`nil` - an empty value.
 
-Each witness is a queue of instructions.
+`Any` - any data type. MUST NOT be `nil`.
 
-In every execution cycle a single instruction gets dequeued and a matching substitution rule gets applied to the stack.
+`Int` - an integer value. We treat the domain of integers as infinite,
+the overflow behaviour or mapping to the actual data types is undefined
+in this spec and should be up to implementation.
 
-In the end, when there are no more instructions left, there MUST be only one item left on the stack.  
+`Hash` - 32 byte value, representing a result of Keccak256 hashing.
 
-### Nodes
-
-There are multiple types of nodes that can be present on the stack.
+`ByteArray` - a byte array of arbitrary size. MUST NOT be empty.
 
-- `hashNode`
-- `branchNode`
-- `accountNode`
-- `valueNode`
-- `extensionNode`
+`()` - an empty array of arbitrary type.
 
-## Substitution rules
+`(Type1 Type1 Type1)` - an array of a type `Type1`.
 
-Here is an example substitution rule. The current instruction is named `INSTR1`.
+### Nodes
 
 ```
-STACK(node1, hash1) STACK(node2, hash2) INSTR1(params) |=> 
-STACK(helper1(node1, node2), helper2(hash1, hash2))
+type Node = HashNode{raw_hash:Hash}
+          | ValueNode{raw_value:ByteArray}
+          | AccountNode{nonce:Int balance:Int storage:nil|Node code:nil|CodeNode|HashNode}
+          | LeafNode{key:ByteArray value:ValueNode|AccountNode}
+          | ExtensionNode{key:ByteArray child:Node}
+          | BranchNode{child0:nil|Node child1:nil|Node child3:nil|Node ... child15:nil|Node}
+          | CodeNode{code:ByteArray}
 ```
 
-This substitution rule replaces 2 stack items `(node1, hash1)` and `(node2, hash2)`
-with a single stack item `(helper1(node1, node2), helper2(hash1, hash2))`.
+### Witness
 
-Where `helper1` and `helper2` are pure functions defined in pseudocode (see the example below).
+Witness MUST have at least 1 element.
 
 ```
-helper1 (value1, value2) {
-    return value1 + value2
+type WitnessHeader = {version:Int}
+type Witness = (Node|Instruction{code:Int parameter:Any...} ...)
```

Comment:
Perhaps we should briefly describe Instruction beforehand.

Reply:
yeah, I was back and forth about it, make sense